### PR TITLE
refactor(api): asyncio smoothie driver

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

A refactor of the smoothie driver to use `AsyncSerial` instead of `SerialCommunication`. This also makes the `Controller` class async and brings us a step closer to not needing the `ThreadManager`.

# Changelog

- Reorganized `opentrons.drivers.smoothie_drivers` package into multiple modules: `simulator, errors, connection, driver, parse_utils, and types`. Most of these contents had been in the `driver_3_0` module.
- Moved `_remove_unwanted_characters` out of `SmoothieDriver` into a `SerialConnection` mixin.
- `SmoothieDriver` uses the asyncio serial connection
- Many `Controller` methods become async and are now awaited by `API` class.
- Split `test_driver.py` tests into unit and integration tests.
- Allow overriding the default timeout on reads and writes in `AsyncSerial` to accomodate Smoothie drivers multiple timeouts.

# Review requests

Besides the unit and integration tests, tested using `SWIFT_SMOKE_FEB_5` using emulation in docker.

# Risk assessment

None. This is being merged into a feature branch.

### CODE REVIEW

File: api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py [hunk 1]
```diff
@@ -2071,10 +1837,10 @@ async def update_firmware(self,  # noqa: C901
                 pass
             else:
                 raise
-
-        if not self.is_connected():
+        #
```
Comment (sfoster1) [inline api/src/opentrons/drivers/smoothie_drivers/driver_3_0.py]: typo?

**Review by sfoster1 — APPROVED** [review]
This is awesome

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py [hunk 2]
```diff
@@ -52,37 +56,41 @@ def __init__(
...
 
-    async def read_until(self, match: bytes) -> bytes:
+    async def read_until(self, match: bytes, timeout: Optional[float] = None) -> bytes:
         """
         Read data until match.
 
         Args:
             match: a sequence of bytes to match
+            timeout: optional timeout in seconds. this is a temporary override
+                of parameter supplied to create
```
Comment (sanni-t) [inline api/src/opentrons/drivers/asyncio/communication/async_serial.py]:
```suggestion
                of parameter supplied to `create`
```

File: api/src/opentrons/drivers/asyncio/communication/async_serial.py [hunk 3]
```diff
@@ -52,37 +56,41 @@ def __init__(
...
 
-    async def write(self, data: bytes) -> None:
+    async def write(self, data: bytes, timeout: Optional[float] = None) -> None:
         """
         Write data
 
         Args:
             data: data to write.
+            timeout: optional timeout in seconds. this is a temporary override
+                of parameter supplied to create
```
Comment (sanni-t) [inline api/src/opentrons/drivers/asyncio/communication/async_serial.py:84]:
```suggestion
                of parameter supplied to `create`
```

File: api/src/opentrons/drivers/smoothie_drivers/__init__.py [hunk 4]
```diff
@@ -1,106 +1,7 @@
...
-
-    @property
-    def axis_bounds(self):
-        position = HOMED_POSITION.copy()
-        position['Y'] = Y_BOUND_OVERRIDE
-        return position
+__all__ = [
+    "SmoothieDriver",
+    "SimulatingDriver"
+]
```
Comment (ahiuchingau) [inline api/src/opentrons/drivers/smoothie_drivers/__init__.py:7]: This is great!! ✨

File: api/src/opentrons/tools/args_handler.py [hunk 5]
```diff
@@ -13,9 +13,9 @@ def root_argparser(description: str = None):
...
 
-def build_driver(
+async def build_driver(
         port: str = None)\
-        -> Tuple[adapters.SynchronousAdapter, SmoothieDriver_3_0_0]:
-    hardware = ThreadManager(API.build_hardware_controller, None, port).sync
-    driver = hardware._backend._smoothie_driver
+        -> Tuple[API, SmoothieDriver]:
+    hardware = await API.build_hardware_controller(port=port)
+    driver = cast(SmoothieDriver, hardware._backend._smoothie_driver)
```
Comment (ahiuchingau) [inline api/src/opentrons/tools/args_handler.py:20]: Out of curiosity, what's the benefit of using `typing.cast` here?

**Review by ahiuchingau — APPROVED** [review]
LGTM! just got a question

File: api/tests/opentrons/drivers/smoothie_drivers/test_driver.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+@pytest.fixture
+async def mock_connection() -> AsyncMock:
+    """The mock SerialConnection."""
+    return AsyncMock(spec=SmoothieConnection)
+
+
+@pytest.fixture
+def sim_gpio() -> SimulatingGPIOCharDev:
+    """The mock GPI."""
```
Comment (sanni-t) [inline api/tests/opentrons/drivers/smoothie_drivers/test_driver.py]:
```suggestion
    """The mock GPIO."""
```

**Review by sanni-t — COMMENTED** [review]
Was just passing by and noticed these
🏃🏻‍♀️

File: api/src/opentrons/tools/args_handler.py (hunk 5 above)
Comment (amitlissack) [reply api/src/opentrons/tools/args_handler.py:20]: It's to make mypy happy. The `_backend._smoothie_driver` member could be a simulating one.

**Review by Laura-Danielle — APPROVED** [review]

**Pull request merged**